src/hotspot/share/opto/escape.cpp

```diff
@@ -401,6 +416,239 @@ bool ConnectionGraph::compute_escape() {
+// details.
+bool ConnectionGraph::can_reduce_this_phi(PointsToNode* var) const {
+  assert(var->escape_state() == PointsToNode::NoEscape, "Phi should be NoEscape");
+
+  PhiNode* phi = var->ideal_node()->as_Phi();
+  const Type* phi_t = _igvn->type(phi);
+  if (phi_t == NULL || phi_t->isa_instptr() == NULL || !phi_t->is_instptr()->klass_is_exact()) {
+    return false;
+  }
+
+  if (!can_reduce_this_phi_inputs(phi) || !can_reduce_this_phi_users(phi)) {
+    return false;
+  }
+
+  NOT_PRODUCT(if (TraceReduceAllocationMerges) { tty->print_cr("Can reduce this Phi during invocation %d: ", _invocation); phi->dump(); })
+  return true;
+}
+
+// This method will create SafePointScalarObjectNode for each combination of
+// scalar replaceable allocation in 'var' and SafePoint node in 'safepoints'.
+// Each SafePointScalarObjectNodes created here may describe multiple scalar
+// replaced objects - check detailed description in SafePointScalarObjectNode
+// class header.
+//
+// This method will set to 'NULL' entries in the Phi that are scalar replaceable.
+void ConnectionGraph::reduce_this_phi_on_safepoints(LocalVarNode* var, Unique_Node_List* safepoints) {
+  assert(var->ideal_node()->is_Phi(), "Not a Phi!");
+  PhiNode* ophi             = var->ideal_node()->as_Phi();
+  Node* minus_one           = _igvn->register_new_node_with_optimizer(ConINode::make(-1));
+  Node* selector            = _igvn->register_new_node_with_optimizer(PhiNode::make(ophi->region(), minus_one, TypeInt::INT));
+  Node* null_ptr            = _igvn->makecon(TypePtr::NULL_PTR);
+  const TypeOopPtr* merge_t = _igvn->type(ophi)->make_oopptr();
+  uint number_of_sr_objects = 0;
+
+  _igvn->hash_delete(ophi);
+
+ // Fill in the 'selector' Phi. If index 'i' of the selector is:
+ // -> a '-1' constant, the i'th input of the original Phi is NSR.
+ // -> a 'x' constant >=0, the i'th input of of original Phi will be SR and the
+ //    info about the scalarized object will be at index x of
+ //    ObjectMergeValue::possible_objects
+ for (uint i = 1; i < ophi->req(); i++) {
+   Node* base          = ophi->in(i);
+   JavaObjectNode* ptn = unique_java_object(base);
+
+   if (ptn != NULL && ptn->scalar_replaceable()) {
+     Node* sr_obj_idx = _igvn->register_new_node_with_optimizer(ConINode::make(number_of_sr_objects));
+     selector->set_req(i, sr_obj_idx);
+     number_of_sr_objects++;
+   }
+ }
+
+ // Update the debug information of all safepoints in turn
+ for (uint spi = 0; spi < safepoints->size(); spi++ ) {
+   Node* call               = safepoints->at(spi);
+   Node* ctrl               = call->in(TypeFunc::Control);
+   Node* memory             = call->in(TypeFunc::Memory);
+   JVMState *jvms           = call->jvms();
+   uint scalar_fields_index = (call->req() - jvms->scloff());
+   int debug_start          = jvms->debug_start();
+   int debug_end            = jvms->debug_end();
+
+   // Keep a copy of the original pointer to NSR objects
+   call->add_req(ophi);
```

Comment:
This a copy of the original pointer to the Phi, which could/should contain at least one SR object as well? The comment implies it's a pointer only to NSR objects.

Reply:
Once the current method return `ophi` will only have NSR inputs. All SR inputs will be set to NULL.